File: geos-gtfv3/geos_gtfv3_interface.c
#include <stdio.h>
#include <time.h>
#include "mpi.h"
#include "geos_gtfv3_interface_py.h"

void geos_gtfv3_interface_c_init(
    MPI_Fint comm_f,
    int npx, int npy, int npz, int ntiles,
    int is, int ie, int js, int je,
    int isd, int ied, int jsd, int jed,
    float bdt, int nq_tot)
{
    MPI_Comm comm_c = MPI_Comm_f2c(comm_f);
    geos_gtfv3_interface_py_init(
        comm_c,
        npx, npy, npz, ntiles,
        is, ie, js, je, isd, ied, jsd, jed, bdt, nq_tot);
}

void geos_gtfv3_interface_c(
    // input
    MPI_Fint comm_f,
    int npx, int npy, int npz, int ntiles,
    int is, int ie, int js, int je,
    int isd, int ied, int jsd, int jed,
    float bdt, int nq_tot, int ng, float ptop, int ks, int layout_1, int layout_2,
    int adiabatic,

    // input/output
    float *u, float *v, float *w, float *delz,
    float *pt, float *delp, float *q,
    float *ps, float *pe, float *pk, float *peln, float *pkz,
    float *phis, float *q_con, float *omga, float *ua, float *va, float *uc, float *vc,

    // input
    const float *ak, const float *bk,

    // input/output
    float *mfx, float *mfy, float *cx, float *cy, float *diss_est)
{

    MPI_Comm comm_c = MPI_Comm_f2c(comm_f);
    /* printf("MPI communicator (F): %d\n", comm_f); */
    /* printf("MPI communicator (C): %d\n", comm_c); */

    // time_t now = time(&now);
    // char buf[20];
    // strftime(buf, sizeof(buf), "%FT%T", localtime(&now));
    // int rank;
    // MPI_Comm_rank(comm_c, &rank);
    // if (rank == 0) printf("C: %s.xxx --calling python interface\n", buf);

    geos_gtfv3_interface_py(
        comm_c,
        npx, npy, npz, ntiles,
        is, ie, js, je,
        isd, ied, jsd, jed,
        bdt, nq_tot, ng, ptop, ks, layout_1, layout_2,
        adiabatic,

        u, v, w, delz,
        pt, delp, q,
        ps, pe, pk, peln, pkz,
        phis, q_con, omga,
        ua, va, uc, vc,

        ak, bk,

        mfx, mfy, cx, cy, diss_est);
}

void geos_gtfv3_interface_c_finalize()
{
    geos_gtfv3_interface_py_finalize();
}

File: geos-gtfv3/geos_gtfv3_interface.f90
module geos_gtfv3_interface_mod

  use iso_c_binding, only: c_int, c_float, c_double

  implicit none

  private
  public :: geos_gtfv3_interface_init_f
  public :: geos_gtfv3_interface_f
  public :: geos_gtfv3_interface_finalize_f

  interface

    subroutine geos_gtfv3_interface_init_f(comm, &
      npx, npy, npz, ntiles, &
      is, ie, js, je, &
      isd, ied, jsd, jed, &
      bdt, nq_tot) bind(c, name='geos_gtfv3_interface_init_c')
      import c_int, c_float, c_double

      implicit none
      integer(kind=c_int), value, intent(in) :: comm
      integer(kind=c_int), value, intent(in) :: npx, npy, npz, ntiles
      integer(kind=c_int), value, intent(in) :: is, ie, js, je
      integer(kind=c_int), value, intent(in) :: isd, ied, jsd, jed
      real(kind=c_float), value, intent(in) :: bdt ! large time step
      integer(kind=c_int), value, intent(in) :: nq_tot ! transported tracers
    end subroutine geos_gtfv3_interface_init_f

     subroutine geos_gtfv3_interface_f( &
          ! Input
          comm, &
          npx, npy, npz, ntiles, &
          is, ie, js, je, &
          isd, ied, jsd, jed, &
          bdt, nq_tot, ng, ptop, ks, layout_1, layout_2, &
          adiabatic, &

          ! Input/Output
          u, v, w, delz, &
          pt, delp, q, &
          ps, pe, pk, peln, pkz, & ! these 5 can be re-computed from delp and ptop
          phis, q_con, omga, ua, va, uc, vc, &

          ! Input
          ak, bk, &

          ! Input/Output
          mfx, mfy, cx, cy, diss_est &
          ) bind(c, name='geos_gtfv3_interface_c')

       import c_int, c_float, c_double

       implicit none

       ! Input
       integer(kind=c_int), value, intent(in) :: comm
       integer(kind=c_int), value, intent(in) :: npx, npy, npz, ntiles
       integer(kind=c_int), value, intent(in) :: is, ie, js, je
       integer(kind=c_int), value, intent(in) :: isd, ied, jsd, jed
       real(kind=c_float), value, intent(in) :: bdt ! large time step
       integer(kind=c_int), value, intent(in) :: nq_tot ! transported tracers
       integer(kind=c_int), value, intent(in) :: ng
       real(kind=c_float), value, intent(in) :: ptop
       integer(kind=c_int), value, intent(in) :: ks
       integer(kind=c_int), value, intent(in) :: layout_1, layout_2
       logical, value, intent(in) :: adiabatic

       ! Input/Output
       real(kind=c_float), dimension(*), intent(inout) :: u, v, w, delz
       real(kind=c_float), dimension(*), intent(inout) :: pt, delp, q
       ! These 5 can be re-computed from delp and ptop
       real(kind=c_float), dimension(*), intent(inout) :: ps, pe, pk, peln, pkz
       real(kind=c_float), dimension(*), intent(inout) :: phis, q_con, omga, ua, va, uc, vc

       ! Input
       real(kind=c_float), dimension(*), intent(in) :: ak, bk

       ! Input/Output
       real(kind=c_float), dimension(*), intent(inout) :: mfx, mfy, cx, cy, diss_est

     end subroutine geos_gtfv3_interface_f

     subroutine geos_gtfv3_interface_finalize_f() bind(c, name='geos_gtfv3_interface_finalize_c')
     end subroutine geos_gtfv3_interface_finalize_f

  end interface

end module geos_gtfv3_interface_mod
